[quant] Add dequantize_per_channel in quantized_decomposed op library

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89272
* #89271
* #89270
* __->__ #89269
* #89268
* #89248

Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

jerryzh168 committed:
[quant] Add dequantize_per_channel in quantized_decomposed op library

Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -122,3 +123,27 @@ def quantize_per_channel(input, scales, zero_points, axis, quant_min, quant_max,
+        # TODO: investigate why
+        # (input[i] - zero_points[i]).to(torch.float32) * scales[i]
+        # failed the test
+        res[i] = (input[i].to(torch.float32) - zero_points[i]) * scales[i]
```
I think it's pretty important for all the quantize/dequantize functions of all flavors to be doing exactly the same thing, regardless of whether they are using quantized tensor or the decomposed representation.

What do you think about having this line of code implemented in c++ with the decomposed signature, and having that used everywhere?

jerryzh168 replied:
yeah I think it makes sense, and ideally fake quant should use the same so that it's exactly matching when people use QAT as well

jerryzh168 replied:
opened an issue in https://github.com/pytorch/pytorch/issues/89303 to track this

jerryzh168 committed:
Update on "[quant] Add dequantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add dequantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_dequantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge -f "failed tests are not relevant to this PR"
